### ISSUE TITLE
TagClosedWrong error when having nested equal formatting elements

### ISSUE THREAD

**Issue by gamezjulian** [AngleSharp/AngleSharp#1052]
# Bug Report

## Prerequisites

- [x] Can you reproduce the problem in a [MWE](https://en.wikipedia.org/wiki/Minimal_working_example)?
- [x] Are you running the latest version of AngleSharp? No, 0.16.1. But it is also replicable in the latest version.
- [x] Did you check the FAQs to see if that helps you?
- [x] Are you reporting to the correct repository? (there are multiple AngleSharp libraries, e.g., `AngleSharp.Css` for CSS support)
- [x] Did you perform a search in the issues?

## Description

Working on a functionallity to sanitize HTML I found out that with some specifc html structure the parser fails with a TagClosedWrong error. However, the html is well formed. The issue seems to be related to the number of nested formatting elements 
withing the provided html document. If the same html element is repeated 3 times (one inside the other), some of them are removed causing that, in later validations, these nodes are detected as not properly closed elements. 

## Steps to Reproduce

1. Make sure you are running a code that has a method attached to the Error handler, like this:
```c
var defaultHtmlParser = new HtmlParser(default, BrowsingContext.New(Configuration.Default.WithCss(new CssParserOptions
            {
                IsIncludingUnknownDeclarations = true,
                IsIncludingUnknownRules = true,
                IsToleratingInvalidSelectors = true,
            })));

            defaultHtmlParser.Error += (sender, e) =>
            {
                var htmlError = e as HtmlErrorEvent;

                if (htmlError != null)
                {
                    this.logger.LogInformation("An error occurred while parsing: {htmlError.Message}", htmlError.Message);
                }
                else
                {
                    this.logger.LogInformation("An error of type {e.Type} occurred while parsing", e.Type);
                }
            };
```
2. Run the ParseDocument method with the following html template
```html
<!doctype html>

<head></head>

<body>
    <div class="jive-rendered-content">
        <p style="margin: 0;">
            <strong style="color: rgba(0, 0, 0, 1); font-size: 14pt; font-family: Calibri;"> <strong
                    style="color: rgba(0, 0, 0, 1); font-size: 14pt; font-family: Calibri;">
                    <strong style="color: rgba(0, 0, 0, 1); font-size: 14pt; font-family: Calibri;">
                        <strong style="color: rgba(0, 0, 0, 1); font-size: 14pt;">
                            <strong style="color: rgba(0, 0, 0, 1); font-size: 14pt; font-family: Calibri;">
                                <strong style="color: rgba(255, 0, 0, 1); font-size: 14pt; font-family: Calibri;">
                                    <strong style="color: rgba(0, 0, 0, 1); font-size: 14pt; font-family: Calibri;">
                                        <strong style="color: rgba(0, 0, 0, 1); font-size: 14pt; font-family: Calibri;">
                                            <strong
                                                style="color: rgba(0, 0, 0, 1); font-size: 14pt; font-family: Calibri;">
                                                ONLY ONE AC TKT IN A PNR </strong>
                                        </strong>
                                    </strong>
                                </strong>
                            </strong>
                        </strong>
                    </strong>
                </strong>
            </strong>
        </p>
    </div>
</body>

</html>

```
4. Check out the error thrown in the error handler.

**Expected behavior:** [What you expected to happen]
The html template should be parsed without errors as it has a valid html structure.

**Actual behavior:** [What actually happened]
The parsing fails with an error saying there was an incorrectly tag closed.

<img width="759" alt="image" src="https://user-images.githubusercontent.com/21286002/176903014-4b898171-7a72-449e-aaa8-3aba0ca5271d.png">

**Environment details:** [OS, .NET Runtime, ...]
Windows, .NET 6.0.

## Possible Solution
If tags are removed to prevent duplicated nested elements, next validations should consider these changes while checking for unclosed elements.

**Comment by driekus77** [comment]
Puzzling bug. Tokenizer looks fine but the formatingElements array seems to react strange to me.

Found suspicious code in HtmlDomBuilderExtensions: 
https://github.com/AngleSharp/AngleSharp/blob/d8bb5245c962eca7d18cbe31b85a803cedb862a9/src/AngleSharp/Html/Parser/HtmlDomBuilderExtensions.cs#L116

```csharp
            for (var i = formatting.Count - 1; i >= 0; i--)
            {
                var format = formatting[i];

                if (format is null)
                {
                    break;
                }

                if (format.NodeName.Is(element.NodeName) &&
                    format.NamespaceUri.Is(element.NamespaceUri) &&
                    format.Attributes.SameAs(element.Attributes) && ++count == 3)
                {
                    formatting.RemoveAt(i);
                    break;
                }
            }
```
When I comment out above part the 1052 unit test succeeds. But now other unit tests are failing...

**Comment by FlorianRappl** [comment]
This might be helpful: https://www.w3.org/TR/2011/WD-html5-20110405/parsing.html#push-onto-the-list-of-active-formatting-elements

**Comment by FlorianRappl** [comment]
Fixed, landed in `devel` (preview of 1.0.0).
